docs/isl-2-0/spec.md

```diff
@@ -0,0 +1,673 @@
+The list of annotation symbols must match the given type.
+
+{% capture sample_code %}
+annotations: { contains: [red, blue] }                   // Annotations must contain "red" and "blue" in any order, and may contain other annotations, such as "yellow"
+annotations: { element: { valid_values: [red, blue] } }  // Only the annotations "red" and "blue" are permitted, but they are not required
+annotations: { container_length: 0 }                     // No annotations are permitted
+annotations: { element: { regex: "\\w+(\\.\\w+)" } }     // Annotations must match the regex
+{% endcapture %}
+{% include example.md title="`annotations` constraint; standard syntax" code=sample_code %}
+
+### Simple Syntax
+Argument may be a list of allowed annotation symbols. 
+The annotations are required to be present on the value if the list is annotated with `required`
+Additional annotations can be prevented by adding the `closed` annotation to the list of valid annotations.
+The list must be annotated with at least one of `closed` or `required`.
+Repeated symbols in the list have no effect; the list should be treated as if it were a set.
+
+While similar to `annotations` in Ion Schema 1.0, it differs in the following ways:
+- `ordered::` annotation is removed
+- at least one of `closed::` or `required::` is required on the list of annotations
+- annotation level `required::` and `optional::` annotations are removed
+
+{% capture sample_code %}
+annotations: required::[red, blue]                // Annotations must contain "red" and "blue" in any order, and may contain other annotations, such as "yellow"
+annotations: closed::required::[red, green, blue] // Annotations must contain exactly "red", "green", and "blue" in any order
+annotations: closed::[red, blue]                  // Only the annotations "red" and "blue" are permitted, but they are not required
+annotations: closed::[]                           // No annotations are permitted
+{% endcapture %}
+{% include example.md title="`annotations` constraint; simplified syntax" code=sample_code %}
+
+Anything that can be represented by the simplified syntax can also be represented by the standard syntax.
+
+<!-- TODO: Add examples of equivalence between standard and simplified syntax -->
+
+## any_of
+
+{% include grammar-element.md productions="any_of" %}
+
+The value must match any of the specified types. The list of types must not be empty.
+
+{% include example.md title="`any_of` constraint" code_file="examples/placeholder.isl" %}
+
+## byte_length
+
+{% include grammar-element.md productions="byte_length" %}
+
+The exact or minimum/maximum number of bytes in a blob or clob.
+Note that this constrains the number of bytes in the input source, which may differ from the number of bytes needed to serialize the blob/clob.
```

Comment:
Not sure I understood this bit. If I have a text clob `{{'''hello'''}}`, there are 15 bytes of input but the clob's content is only 5 bytes long. `byte_length` constrains the 5, right? Not the 15?

Reply:
Correct. `byte_length` constrains the value rather than the encoded representation. I'll rework this to make it more clear.